concepts/functions/about.md

```diff
@@ -1,3 +1,176 @@
+```rust
+fn do_nothing() {}
+
+```
+
+The empty parentheses are a placeholder for zero or more parameters. A parameter is the definition of a value to be passed into the function.
+Even if there are no parameters the parentheses are still required in the function definition. The actual value passed to a parameter is called
+an argument.
+
+The combination of parameters and return value is known as the function's signature. A function's signature requires that each parameter must
+have its type annotated.
+
+The curly braces enclose the body of the function. Inside the braces are where the function's statements and expressions are placed. Even if the
+function contains no statements or expressions the curly braces are still required for the function definition. Of course there is no point to
+having an empty function.
+
+A function could contain a simple statement. In the following example the function takes in one `i32` argument and prints it
+
+```rust
+fn print_integer(value: i32) {
+    println!("{:?}", value);
+}
+```
+
+Note that the parameter needs to be given a name (`value` in this case) and be annotated with its type.
+
+A function can also return an expression. In the following example the function takes in one `i32` argument and returns it doubled
+
+```rust
+fn double_integer(value: i32) -> i32 {
+    value * 2
+}
+```
+
+The `-> i32` indicates that the function returns an `i32`. Unlike parameters, the returned value is not named. Note that the last expression
+in a function can be returned without the `return` keyword and a semicolon. `return` _is_ needed to return early from a function, like so
+
+```rust
+fn long_function() -> i32 {
+    let some_condition = false;
+    // code snipped
+    if (some_condition) {
+        return 0;
+    }
+    // more code snipped
+    42
+}
+```
+
+A function has its own scope which borrows or takes ownership of arguments passed to it. In this example the last line in `main` causes a
+compile error
+
+```rust
+fn double_integer(value: Box<i32>) -> i32 {
+    *value * 2
+}
+
+pub fn main() {
+    let num = Box::new(42);
+    let num_doubled = double_integer(num); // value moved here
+    println!("{:?}", num_doubled);
+    println!("{:?}", num); // borrow of moved value: `num` value borrowed here after move
+    // move occurs because `num` has type `Box<i32>`, which does not implement the `Copy` trait
+}
+```
+
+`main`, which owns the boxed `num` value, moves `num` as the argument for `value` into `double_integer`, which takes ownership of it. At the
+end of `double_integer`, `value` goes out of scope and is dropped, so `num` is no longer valid for the rest of the `main` function.
+
+Note that `fn main` is prefaced by `pub`, but `fn double_integer` is not. A `pub fn` is public, meaning it can be accessed from a different
+module. Functions defined without `pub` are private and can only be accessed from the current module. `pub(crate) fn` makes a function
+accessible only within the current crate.
```

Comment:
This information is good but belongs to a different concept. Not sure which precisely; it could go into `scope`, but I can also see an argument for putting it into `module-system`, `visibility`, or something else. Either way we don't have to teach it here.

Also, it's somewhat odd to me to see `pub fn main`, because the `pub` is unnecessary; `clippy` doesn't say anything about it, but in most projects I've seen, it's just `fn main`.

Reply:
I'd propose `visibility` since we do not have a `concepts/module-system`.

Though, should we add a new concept `module-system`?

Reply:
I see articles every once in a while whose premise is "Rust's module system is confusing; here's how it works." Anecdotally, I feel like we should explicitly teach the module system _somewhere_. I don't have strong feelings about where.

`scope`, `visibility`, and `modules` are very closely coupled concepts. Perhaps for now we should just try to teach all of them in `scope_and_visibility` or something, and then just split things out if that concept gets too big.